nevow/appserver.py

```diff
@@ -29,6 +31,81 @@
 from nevow import stan
 
 
+
+class _DictHeaders(MutableMapping):
+    """
+    A C{dict}-like wrapper around L{Headers} to provide backwards compatibility
+    for L{twisted.web.http.Request.received_headers} and
+    L{twisted.web.http.Request.headers} which used to be plain C{dict}
+    instances.
+
+    @type _headers: L{Headers}
+    @ivar _headers: The real header storage object.
+    """
+    def __init__(self, headers):
+        self._headers = headers
+
+
+    def __getitem__(self, key):
+        """
+        Return the last value for header of C{key}.
+        """
+        if self._headers.hasHeader(key):
+            return self._headers.getRawHeaders(key)[-1]
+        raise KeyError(key)
+
+
+    def __setitem__(self, key, value):
+        """
+        Set the given header.
+        """
+        self._headers.setRawHeaders(key, [value])
+
+
+    def __delitem__(self, key):
+        """
+        Delete the given header.
+        """
+        if self._headers.hasHeader(key):
+            self._headers.removeHeader(key)
+        else:
+            raise KeyError(key)
+
+
+    def __iter__(self):
+        """
+        Return an iterator of the lowercase name of each header present.
+        """
+        for k, v in self._headers.getAllRawHeaders():
+            yield k.lower()
+
+
+    def __len__(self):
+        """
+        Return the number of distinct headers present.
+        """
+        # XXX Too many _
```

Comment:
This is copied verbatim from Twisted; my plan is to remove the code again once this functionality is removed from Nevow for good, rather than maintain it, but there should be a ticket about this so I'll file one.